Implemented qemu guest agent

lukas016 opened the pull request:
# Proposed Changes

- Extend `api.Machine` spec and status
- Add new argument
- Add new logic into server and reconciler

Fixes #165

lukas016 committed:
implement qemu guest agent

- extend api.Machine spec and status
- add new argument
- add new logic into server and reconciler

lukasfrank commented on pkg/api/machine.go:
```diff
@@ -28,13 +28,16 @@ type MachineSpec struct {
 
 	ShutdownAt time.Time `json:"shutdownAt,omitempty"`
+
+	QemuGuestAgentEnable bool
```
```suggestion
	QemuGuestAgent GuestAgent
```

Can we maybe update the spec? 
```
type GuestAgent string 
const (
	EnableGuestAgent GuestAgent = "Enable"
	DisableGuestAgent GuestAgent = "Disable"
)
```

lukas016 replied:
What are benefits? Because we have some approach for Hugepages. Do will we need more values as true/false? Compare of strings can has higher impact to performance too.

lukasfrank replied:
These are the k8s api conventions. The general reasoning is that the string types can be extended if needed.

lukas016 replied:
we can go this way, but i prefer lower case letters. It is simpler for comparison and i don't have to do conversation or support multiple variant.

lukas016 replied:
i  will implement it as you suggest for keep the same format. Conversion to lower case we can discuss next time.

lukas016 replied:
fixed

lukasfrank commented on pkg/api/machine.go:
```diff
@@ -28,13 +28,16 @@ type MachineSpec struct {
 	NetworkInterfaceStatus []NetworkInterfaceStatus `json:"networkInterfaceStatus"`
 	State                  MachineState             `json:"state"`
 	ImageRef               string                   `json:"imageRef"`
+	QemuGuestAgent         *AgentStatus             `json:"qemuGuestAgent,omitempty"`
```
In general: Do we need `Qemu` in `QemuGuestAgent` ? Isn't that implementation detail?

lukas016 replied:
Theoretically you can have multiple type of agents (https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.3/html/virtual_machine_management_guide/installing_guest_agents_and_drivers_linux#doc-wrapper). But we support qemu currently so i wanted do it extendable for future.

lukas016 replied:
but we can do it with extend agentStatus, so yes we can rename it to GuestAgent.

lukasfrank replied:
I think it's a good idea to put the type of the `GuestAgent` into the status.

lukas016 replied:
fixed

lukas016 committed:
use specific type for guest agent

lukas016 committed:
set guest agent for tests

lukasfrank commented on provider/server/server.go:
```diff
@@ -68,9 +72,24 @@ func setOptionsDefaults(o *Options) {
+	return api.GuestAgent(availableGuestAgents[index]), nil
+}
+
 func New(opts Options) (*Server, error) {
 	setOptionsDefaults(&opts)
 
+	guestAgent, err := convertGuestAgentFromString(opts.GuestAgent)
+	if err != nil {
+		return nil, err
+	}
+
```
wdyt about passing the `GuestAgent` here and to the validation outside of the server?

lukas016 replied:
GuestAgent is currently used in server only. If you want to implement conversion somewhere else, it can be do, but we can follow my suggestion with implement pflag.Value interface because in other case we will end with multiple convert functions this way and implementation of pflag.Value is fast failure.

lukas016 replied:
if you don't like use API type for that. We can move GuestAgent into new package. Maybe something as features/guestagent ? And it can be import into provider/cmd/app and pkg/api as independent package and we can without big problems implement pflag.Value interface

lukas016 replied:
reimplemented as pflag.Value interface:

```
Error: invalid argument "Test" for "--guest-agent-type" flag: unsupported option Test
Usage:
  libvirt-provider [flags]

Flags:
      --address string                             Address to listen on. (default "/var/run/iri-machinebroker.sock")
      --apinet-kubeconfig string                   Path to the kubeconfig file for the apinet-cluster.
      --apinet-node-name string                    APInet node name
```

lukas016 replied:
fixed

lukasfrank commented on pkg/api/machine.go:
```diff
@@ -107,3 +121,8 @@ const (
 )
+
+type AgentStatus struct {
+	Type GuestAgent `json:"type"`
```
Does this bring us any benefit having this here again? If no: I prefer to have the `AgentStatus` as a pointer in the `MachineStatus`

lukas016 replied:
No significant benefits, we can say you don't need handle nil for GuestAgent in status in future. I can revert it.

lukas016 replied:
fixed

lukasfrank replied:
```suggestion
```

lukas016 commented on provider/server/server.go:
```diff
@@ -68,9 +72,24 @@ func setOptionsDefaults(o *Options) {
 }
 
+func convertGuestAgentFromString(guestAgent string) (api.GuestAgent, error) {
+	availableGuestAgents := api.GuestAgentAvailable()
```
i had switch as first solution but you have to handle options in 2 packages (api and here). I don't like it. It can be harder handle it in future and this function is calling just one time during start, so i don't see big benefits in performance improvement with switch.

lukas016 replied:
GuestAgentAvailable was moved

lukas016 commented on provider/server/machine_create.go:
```diff
@@ -89,6 +89,8 @@ func (s *Server) createMachineFromIRIMachine(ctx context.Context, log logr.Logge
 		machine.Spec.Image = &iriMachine.Spec.Image.Image
 	}
 
+	machine.Spec.GuestAgent = s.guestAgent
```
fixed

lukas016 committed:
create guest agent option type for argument

- improve conversion of argument into internal type
- implement fast failure for parsing argument
- improve set GuestAgent in api.Machine.Spec

lukas016 committed:
implement guest agent shutdown mode

so-sahu commented on provider/cmd/app/option_types.go:
```diff
@@ -0,0 +1,48 @@
+	return api.GuestAgent(*g)
+}
+
+func GuestAgentOptionAvailable() []string {
```
Scope of this function can be reduced to package level, as it's only being used within this package.

lukas016 replied:
fixed

lukas016 committed:
apply review comments

lukas016 commented on pkg/api/machine.go:
```diff
@@ -28,13 +28,23 @@ type MachineSpec struct {
 	NetworkInterfaceStatus []NetworkInterfaceStatus `json:"networkInterfaceStatus"`
 	State                  MachineState             `json:"state"`
 	ImageRef               string                   `json:"imageRef"`
+	GuestAgent             *GuestAgentStatus        `json:"guestAgent,omitempty"`
```
fixed

so-sahu reviewed (dismissed):
LGTM

hardikdr commented on pkg/controllers/machine_controller.go:
```diff
@@ -806,6 +811,38 @@ func (r *MachineReconciler) setTCMallocPath(domain *libvirtxml.Domain) error {
 }
 
+func (r *MachineReconciler) setGuestAgent(machine *api.Machine, domainDesc *libvirtxml.Domain) {
+	if machine.Spec.GuestAgent == api.GuestAgentNone {
```
Can we have this change outside, so that we know code doesn't enter at all if it's disabled? Just a cosmetic suggestion.

lukas016 replied:
how do you mean that? I prefer have all logic in function. It is logical block, so you have to just delete this function and calls to it and you don't need study code what else you have to delete.

hardikdr replied:
I would usually prefer to keep the optional features behind the `if`, then we know code-flow never enters into this method when flag is set, and main flow is safe-guarded. 
In this case, it's an optional suggestion, feel free to resolve it otherwise. 
```golang
if conditionMet {
   setGuestAgent()
}
```

lukas016 replied:
fixed

hardikdr reviewed (commented):
Thanks @lukas016 for the PR, just a minor suggestion, looks good otherwise.

lukas016 committed:
check GuestAgent before set

lukas016 committed:
Merge branch 'main' into 165-qemu-guest-agent-support

lukas016 committed:
Merge branch 'main' into 165-qemu-guest-agent-support

lukas016 committed:
Merge branch 'main' into 165-qemu-guest-agent-support

lukasfrank commented on pkg/controllers/machine_controller.go:
```diff
@@ -790,6 +797,34 @@ func (r *MachineReconciler) setTCMallocPath(domain *libvirtxml.Domain) error {
+	}
+
+	domainDesc.Devices.Channels = append(domainDesc.Devices.Channels, agent)
+	machine.Status.GuestAgentStatus = &api.GuestAgentStatus{Type: machine.Spec.GuestAgent, Addr: "unix://" + socketPath}
```
We do not use this property 
```suggestion
	machine.Status.GuestAgentStatus = &api.GuestAgentStatus{ Addr: "unix://" + socketPath}
```

lukas016 replied:
it is information for user. Do you really want to delete it?

lukas016 replied:
wait, i used status.Type in getShutdownMode

lukas016 replied:
fixed

lukasfrank commented on pkg/controllers/machine_controller.go:
```diff
@@ -888,3 +923,11 @@ func machineDomain(machineID string) libvirt.Domain {
 	}
 }
+
+func getShutdownMode(agent *api.GuestAgentStatus) libvirt.DomainShutdownFlagValues {
```
What do you think about inlining this few lines?

lukas016 replied:
theoretically, i prefer more function for simpler extending but it this case no problem.
If you want to simplify it, we can use automode and libvirt will choose best option alone, but i am afraid of unexpected behavior in this case and logic will depend on another component and behavior can change base on version what can bring more problem for us. So i don't recommend use automode.

lukas016 replied:
fixed

lukas016 commented on pkg/controllers/machine_controller.go:
```diff
@@ -678,6 +681,10 @@ func (r *MachineReconciler) domainFor(
 		return nil, nil, nil, err
 	}
 
+	if machine.Spec.GuestAgent != api.GuestAgentNone {
```
ok logic is stable because it is set in call CreateMachine.

lukas016 committed:
remove guest type from status

lukas016 committed:
Merge branch 'main' into 165-qemu-guest-agent-support
